tests/hierarchy_driver.cc

```diff
@@ -174,8 +174,7 @@ int main(int argc, char *argv[])
 {
   namespace boost_po = boost::program_options;
 
-  MPI_Init(&argc, &argv);
-  dealii::MultithreadInfo::set_thread_limit(1);
+  dealii::Utilities::MPI::MPI_InitFinalize mpi_init(argc, argv);
```

Comment:
For requesting explicit thread support we need to call `MPI_Init_thread` instead which `dealii::Utilities::MPI::MPI_InitFinalize` does (apart from initializing Zoltan and emptying memory pools).

`MPI_Init` seems to not allow calling `MPI` functions from multiple threads and hence behaves similar to `MPI_Init_thread` with `MPI_THREAD_SINGLE` (or possibly `MPI_THREAD_FUNNELED`).

In the end, I was just seeing `MPI` errors when calling `MPI` functions (inside the `MatrixFree` constructior to be precise) when running multithreaded. Using `MPI_Init` instead of `MPI_Init_thread` or `dealii::Utilities::MPI::MPI_InitFinalize` was likely just an oversight.

Reply:
Updated link to deal.II code